### PR TITLE
Document that hooks must not raise / propagate exceptions.

### LINKED ISSUE #49: Document exception semantics for hooks
In particular, the `_before_commit` hooks are not expected to raise exceptions, and if they do so, leave the transaction in an ambiguous state (`commit` not called, hooks not cleared, but one or more hooks may have completed).

See: https://github.com/Pylons/pyramid_tm/issues/63

### PR DESCRIPTION
**Pull request by tseaver**
Closes #49.

### CODE REVIEW

File: docs/hooks.rst [hunk 1]
```diff
@@ -1,6 +1,20 @@
 Hooking the Transaction Machinery
 =================================
 
+The :mod:`transaction` machinery allows application developers to register
+two different groups of callbacks to be called, one group before the
```
Comment (jamadden) [inline docs/hooks.rst]: Extra `the`.

File: docs/hooks.rst [hunk 2]
```diff
@@ -1,6 +1,20 @@
...
+
+These hooks are **not** designed to be used as replacements for the
+two-phase commit machinery defined by a resource manager (see
+:doc:`resourcemanager`).  In particular, hook functions **must not** raise
+or propagate exceptions.
+
+.. warning::
+
+   Hook functions which *do* raise or propagate exceptions will leave the
+   application in an undefined state.
```
Comment (jamadden) [inline docs/hooks.rst]: "Application" seems oddly non-specific here. It's really the transaction and its resource managers that's in an undefined state, isn't it? By extension the state of the application could be said to be undefined, I suppose.

Comment (tseaver) [inline docs/hooks.rst]: The transaction **itself** is actually undamaged by misbehaving hooks:

- Exceptions raised from a before-commit hook don't prevent a subsequent attempt to commit the same transaction.
- Exceptions raised from an after-commit hook are logged and swallowed.

What is broken is the application's expectations about the transaction:  e.g. not all hooks get called, etc.  The point is that it is undefined behavior.

File: docs/hooks.rst (hunk 1 above)
Comment (tseaver) [inline docs/hooks.rst]: Thanks!

File: docs/hooks.rst (hunk 2 above)
Comment (jamadden) [inline docs/hooks.rst]: Agreed it’s undefined, and your clarification helps my understanding. Thanks for the elucidation.

**Pull request merged**